exotations/solvers/exotica_ddp_solver/src/analytic_ddp_solver.cpp

```diff
@@ -0,0 +1,112 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <exotica_ddp_solver/analytic_ddp_solver.h>
+
+REGISTER_MOTIONSOLVER_TYPE("AnalyticDDPSolver", exotica::AnalyticDDPSolver)
+
+namespace exotica
+{
+AnalyticDDPSolver::AnalyticDDPSolver() = default;
+AnalyticDDPSolver::~AnalyticDDPSolver() = default;
+
+void AnalyticDDPSolver::Instantiate(const AnalyticDDPSolverInitializer& init)
+{
+    parameters_ = init;
+    base_parameters_ = parameters_;
+}
+
+void AnalyticDDPSolver::BackwardPass()
+{
+    constexpr double min_clamp_ = -1e10;
+    constexpr double max_clamp_ = 1e10;
+    const int T = prob_->get_T();
+    const int NX = prob_->get_num_positions();
+    const int NU = prob_->get_num_controls();
+    const int N = NX + prob_->get_num_velocities();
+    const double dt_squared = dynamics_solver_->get_dt() * dynamics_solver_->get_dt();
+
+    Eigen::MatrixXd Qx, Qu, Qxx, Quu, Qux, Quu_inv, Vxx;
+    Eigen::VectorXd Vx;
+
+    // V = prob_->GetStateCost(T - 1);
+    Vx = prob_->GetStateCostJacobian(T - 1);
+    Vxx = prob_->GetStateCostHessian(T - 1);
+
+    // concatenation axis for tensor products
+    Eigen::array<Eigen::IndexPair<int>, 1> dims = {Eigen::IndexPair<int>(1, 0)};
+
+    for (int t = T - 2; t > 0; t--)
+    {
+        Eigen::VectorXd x = prob_->get_X(t), u = prob_->get_U(t);
+        Eigen::MatrixXd fx = dynamics_solver_->fx(x, u),
+                        fu = dynamics_solver_->fu(x, u);
+
+        fx = fx * dynamics_solver_->get_dt() + Eigen::MatrixXd::Identity(fx.rows(), fx.cols());
+        fu = fu * dynamics_solver_->get_dt();
+
+        // Q = prob_->GetStateCost(t) + prob_->GetControlCost(t) + V; // l + v
+        Qx = prob_->GetStateCostJacobian(t) + fx.transpose() * Vx;  // lx + fx @ Vx
+        Qu = prob_->GetControlCostJacobian(t) + fu.transpose() * Vx;
+        Qxx = prob_->GetStateCostHessian(t) + fx.transpose() * Vxx * fx;
+        Quu = prob_->GetControlCostHessian() + fu.transpose() * Vxx * fu;
+
+        // NOTE: Qux = Qxu for all robotics systems I have seen
+        //  this might need to be changed later on
+        // WARNING_NAMED("ControlLimitedDDPSolver", "Assuming fux = fxu. Make sure this is valid for your robot!");
+        Qux = prob_->GetStateControlCostHessian() + fu.transpose() * Vxx * fx;
+
+        if (parameters_.UseSecondOrderDynamics)
```

Comment:
I am trying to think whether it makes sense to have a full if else - and not just add, but the whole equations on their own. Reason being that Eigen/compiler may do some optimisation and it's also easier to read. I.e., if (secondorderdynamics), the whole equations for Qxx, Quu, Qux, else - reduced equations as above.

Reply:
Did that.